proposals/20230316-oci-content-source.md

```diff
@@ -0,0 +1,284 @@
+#### Non-functional
+- TODO
+
+### Implementation Details
+At the time of authoring this proposal, the OCI v1.1 image specification is still in release candidate state, and there is a bit of flux with regard to artifacts. This proposal will take into account the most up-to-date state of the image specification and may evolve as changes in the OCI specifications evolve.
+
+#### Artifact Types
+Image manifests will be differentiated based on the `artifactType` field in the image manifest. The following media types will be used.
+- Component: "application/vnd.wasm.component.v1"
+- Interface: "application/vnd.wasm.component.interface.v1"
+- Bundled Component: "application/vnd.wasm.component.bundled.v1"
+
+#### Image Manifest for Components
+The following is an example image manifest for a component containing a configuration structure, a layer containing the `my-component.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:5587da2246a78f08c447bff2ac91ee5c2b57be2f2a15244b5e618ac0be626885",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:b36aa5d0111a488937361fdb35432510d50675a11d566c5d9e82a147fb9ff552",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Interface Components
+The following is an example image manifest for a component containing a configuration structure and a layer containing the `my-component-interface.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.interface.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:dcf07c6bb395d6e1d40505b77e70af04f5fae0d54c9573fd379c1e7355a18cf3",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component-interface"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Bundled Components
```

Comment:
If we take "bundled component" to mean "a component with its component-dependencies replaced with inline components" (as defined above), then this artifact type is not a bundled component -- it's a component combined with environment variable configuration and static assets: that's not a component at all, it's something bigger that *contains* a component, so think we should give it a different name.  Because this bigger thing is no longer composable the way components are composable, I'd suggest calling it a "wasm app" (or something that indicates that it's the final product that you can deploy), and not including "component" in the name at all (that's an impl detail of the app, and the app could just as well contain a core wasm module as, I believe, it's doing today).